File: decky/py_modules/decky_my_rig/versioning.py
from __future__ import annotations

import re
from enum import StrEnum


_VERSION = re.compile(r"^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)$")


class VersionRelation(StrEnum):
    COMPATIBLE = "compatible"
    UPDATE_HOST = "update_host"
    UPDATE_PLUGIN = "update_plugin"
    INCOMPATIBLE = "incompatible"
    UNKNOWN = "unknown"


def compare_versions(plugin_version: str, host_version: str) -> VersionRelation:
    plugin = _VERSION.fullmatch(plugin_version)
    host = _VERSION.fullmatch(host_version)
    if plugin is None or host is None:
        return VersionRelation.UNKNOWN
    plugin_parts = tuple(map(int, plugin.groups()))
    host_parts = tuple(map(int, host.groups()))
    if plugin_parts[0] != host_parts[0]:
        return VersionRelation.INCOMPATIBLE
    if plugin_parts[1] > host_parts[1]:
        return VersionRelation.UPDATE_HOST
    if plugin_parts[1] < host_parts[1]:
        return VersionRelation.UPDATE_PLUGIN
    return VersionRelation.COMPATIBLE

File: host/src/versioning.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum VersionRelation {
    Compatible,
    UpdateHost,
    UpdatePlugin,
    Incompatible,
    Unknown,
}

fn parse(value: &str) -> Option<[u64; 3]> {
    let mut parts = value.split('.');
    let parsed: [u64; 3] = [
        parts.next()?.parse().ok()?,
        parts.next()?.parse().ok()?,
        parts.next()?.parse().ok()?,
    ];
    if parts.next().is_some() || parsed.map(|part| part.to_string()).join(".") != value {
        return None;
    }
    Some(parsed)
}

pub fn compare(plugin: &str, host: &str) -> VersionRelation {
    let (Some(plugin), Some(host)) = (parse(plugin), parse(host)) else {
        return VersionRelation::Unknown;
    };
    if plugin[0] != host[0] {
        VersionRelation::Incompatible
    } else if plugin[1] > host[1] {
        VersionRelation::UpdateHost
    } else if plugin[1] < host[1] {
        VersionRelation::UpdatePlugin
    } else {
        VersionRelation::Compatible
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn only_patch_differences_are_silent() {
        assert_eq!(compare("1.2.9", "1.2.0"), VersionRelation::Compatible);
        assert_eq!(compare("1.3.0", "1.2.9"), VersionRelation::UpdateHost);
        assert_eq!(compare("1.2.0", "1.3.0"), VersionRelation::UpdatePlugin);
        assert_eq!(compare("2.0.0", "1.9.0"), VersionRelation::Incompatible);
        assert_eq!(compare("v1.2.3", "1.2.3"), VersionRelation::Unknown);
    }
}
